subprojects/dependency-lock/src/main/java/org/gradle/internal/dependencylock/model/DependencyLock.java

```diff
@@ -0,0 +1,74 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.dependencylock.model;
+
+import org.gradle.api.artifacts.ModuleIdentifier;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+public class DependencyLock {
+
+    private final SortedMap<String, SortedMap<String, LinkedHashMap<ModuleIdentifier, List<DependencyVersion>>>> projectsMapping = new TreeMap<String,  SortedMap<String, LinkedHashMap<ModuleIdentifier, List<DependencyVersion>>>>();
+
+    public void addDependency(String projectPath, String configurationName, ModuleIdentifier moduleIdentifier, DependencyVersion dependencyVersion) {
+        if (projectsMapping.containsKey(projectPath)) {
+            SortedMap<String, LinkedHashMap<ModuleIdentifier, List<DependencyVersion>>> project = projectsMapping.get(projectPath);
+            LinkedHashMap<ModuleIdentifier, List<DependencyVersion>> modulesMapping = createAndGetModulesMapping(project, configurationName);
+
+            if (!modulesMapping.containsKey(moduleIdentifier)) {
+                startCapturingDependenciesForModule(modulesMapping, moduleIdentifier, dependencyVersion);
+            } else {
+                modulesMapping.get(moduleIdentifier).add(dependencyVersion);
+            }
+        } else {
+            SortedMap<String, LinkedHashMap<ModuleIdentifier, List<DependencyVersion>>> configurationMapping = createNewConfigurationMapping(configurationName);
+            LinkedHashMap<ModuleIdentifier, List<DependencyVersion>> modulesMapping = new LinkedHashMap<ModuleIdentifier, List<DependencyVersion>>();
+            startCapturingDependenciesForModule(modulesMapping, moduleIdentifier, dependencyVersion);
+            configurationMapping.put(configurationName, modulesMapping);
+            projectsMapping.put(projectPath, configurationMapping);
+        }
+    }
+
+    private LinkedHashMap<ModuleIdentifier, List<DependencyVersion>> createAndGetModulesMapping(SortedMap<String, LinkedHashMap<ModuleIdentifier, List<DependencyVersion>>> project, String configurationName) {
+        if (!project.containsKey(configurationName)) {
+            project.put(configurationName, new LinkedHashMap<ModuleIdentifier, List<DependencyVersion>>());
+        }
+
+        return project.get(configurationName);
+    }
+
+    private void startCapturingDependenciesForModule(LinkedHashMap<ModuleIdentifier, List<DependencyVersion>> modulesMapping, ModuleIdentifier moduleIdentifier, DependencyVersion dependencyVersion) {
```

Comment:
❌ How can there be multiple versions for a module in a given configuration?

Reply:
See the discussion here: https://github.com/gradle/gradle/pull/3302#discussion_r147806640.

Reply:
I don't understand this. At the end of the day, one version wins and is put into the graph. I don't think the lock file should have an entry per selector, but an entry per module identifier. @bigdaz am I missing something here?

Reply:
No. _If_ we are persisting the _requestedVersion_ in our lock file, then we are locking _edges_ not _nodes_. My comment was that if we are locking edges, then we need to persist all of them.

This is why I was asking for more context on how the lock file would be used. It seemed odd that we store the 'requestedVersion' at all.

Reply:
Agreed, we don't need the requested version. The lock file specifies the version per module, not per selector. That simplifies things a bit more.

Reply:
We'll probably need to store the target module as well in addition to the requested version. That would be the case for a dependency resolve rule where we replace a module with another module e.g. `org.gradle:test -> other.company:foo:1.5`.